### DISCUSSION TITLE
Dynamic Table Name， a lot of tables with the same structure, it is not appropriate to define many structures with macros

### DISCUSSION THREAD

**Discussion by langzime** [SeaQL/sea-orm#1202] [Ideas]

**Comment by billy1624** [comment]
Hey @langzime, is this a follow up question of
- https://github.com/SeaQL/sea-orm/issues/782

**Comment by billy1624** [comment]
Yeah, this is the tradeoff. Defining the repeated structs over and over again OR defining repeated struct with macro. Define with macro will improve the readability and avoid repeated code but it's harder to develop upon as not all IDEs are able to expand the macros.

**Comment by langzime** [comment]
Yes, i tried it. A lot of macros cause ide very laggy and a lot of extra code

**Comment by billy1624** [comment]
From my experience, it's just a one-time long compilation as 200+ entities were added at once. Once it's compiled it will never be recompiled on the same machine given the fact that entity file is unchanged.

**Comment by langzime** [comment]
I have 500+ ...

**Comment by langzime** [comment]
In cases like this, I currently use both mysql_async

**Comment by billy1624** [comment]
I think there are three options:
1. Fork SeaORM and change this part, making it to be provided in runtime: https://github.com/SeaQL/sea-orm/blob/1ea8b457bf9f66988a45696c15933a6a756854ed/src/tests_cfg/cake_expanded.rs#L7-L11
2. Construct SQL statement with SeaQuery then pass it to SeaORM for execution: https://github.com/SeaQL/starfish-ql/blob/459a71609785f74739d4b5ba9a2983af0cedfae6/starfish/starfish-core/src/query/mod.rs#L290-L303

**Reply by canxin121** [reply]
[dynamic String Entity name](https://github.com/canxin121/sea-orm/commit/1ce8ac3ce1de9db4108ab3aaff6bc79db1de9e1f) I tried to remove IdenStatic of EntityName, and use Iden instead. But this related to much more codes than i expected

**Reply by theelderbeever** [reply]
There is a working example in one of the other threads

https://github.com/SeaQL/sea-orm/discussions/1202#discussioncomment-9241634

**Comment by langzime** [comment]
Thank you very much,  I'll try it sometime.
Do you have plans to support this feature?

**Comment by billy1624** [comment]
> Do you have plans to support this feature?

It's not on our plan. But if you have an good proposal we're willing to consider :D

**Comment by tyt2y3** [comment]
Actually, it's possible to have an Entity that has a state, so your Entity is no longer a unit struct, but a `struct Entity(String)`.

**Comment by tyt2y3** [comment]
Although I am not sure are there other blockers to this approach. But yeah at least it was designed that way, the `table_name` takes a `self`

**Comment by billy1624** [comment]
From `table_name` perspective, it's designed with runtime in mind. https://github.com/SeaQL/sea-orm/blob/6bc76698cc54aea1eb54cbd82d1d5af096432079/src/entity/base_entity.rs#L23-L24

However, when constructing query. It doesn't take `&self`. So, it's solely rely on the `Entity` defined in compile-time.

Also, `EntityName` must be `Copy` but `String` isn't. So, `pub struct Entity(pub String)` wouldn't compile.

https://github.com/SeaQL/sea-orm/blob/6bc76698cc54aea1eb54cbd82d1d5af096432079/src/entity/base_entity.rs#L10-L24

**Comment by tyt2y3** [comment]
Then probably a lazy_static with `Entity(&'static str)` would work

**Comment by pedrettin** [comment]
This issue is blocking us from using sea ORM when the table_name is set at runtime. We create many SQL tables with the same struct dynamically and we don't know the names of the tables before hand.

Setting a variable in Entity won't work as the `find()` method is called statically on the Entity object, so creating a new entity object with the right "table_name" isn't useful.

**Comment by tyt2y3** [comment]
Then probably your best bet for now is using [SeaQuery](https://github.com/SeaQL/sea-query/), everything is dynamic there.

**Comment by penso** [comment]
Got a similar issue where I'd like to set the `table_name` dynamically with a prefix based on a more global app context, I'll have multiple copies of the same tables. But it also means ideally to be able to run migrations multiple times for all the table names.

**Comment by theelderbeever** [comment]
Has anyone figured out anything on this? We have the situation where we have a set of tables that are individually generated at runtime with a predetermined naming pattern. I looked into SeaQuery however, `Query::select().from(<IntoTableRef>)` is super restrictive. Is there anyway to just pass a String at runtime for the table that is supposed to be queried?

**Reply by penso** [reply]
Would love to know, badly need this soon...

**Reply by theelderbeever** [reply]
@penso So far I have hacked together this example

```rust
use std::convert::From;
use std::fmt;

use sea_query::*;

struct TableName(String);

impl TableName {
    fn new(s: &str) -> Self {
        Self(String::from(s))
    }
}

impl From<TableName> for SeaRc<dyn Iden> {
    fn from(value: TableName) -> Self {
        SeaRc::new(value)
    }
}

impl Iden for TableName {
    fn prepare(&self, s: &mut dyn fmt::Write, q: Quote) {
        write!(s, "{}", self.quoted(q)).unwrap();
    }

    fn quoted(&self, q: Quote) -> String {
        format!("{}{}{}", q.left(), self.0, q.right())
    }

    fn to_string(&self) -> String {
        self.0.clone()
    }

    fn unquoted(&self, s: &mut dyn fmt::Write) {
        let _ = s.write_str(&self.0);
    }
}

fn main() {
    let select = Query::select()
        .expr(Expr::cust("SUM(1)"))
        .from(TableRef::Table(TableName::new("hello").into()))
        .build(PostgresQueryBuilder);
    println!("{}", select.0);
}
```

which prints

`SELECT SUM(1) FROM "hello"`

**Reply by theelderbeever** [reply]
And then my plan is to define the enums for the columns and such and just do query building. Bear in mind that my tables aren't created via migrations and are instead dynamically generated at runtime as the application sees new data.

**Reply by theelderbeever** [reply]
@penso actually this seems even easier

```rust
use sea_query::*;

pub enum Metrics {
    TableName(String),
    Timestamp,
    MetricId,
    Account,
    IntervalMs,
    Counter,
    Gauge,
    MetricTagIds,
}

// Mapping between Enum variant and its corresponding string value
impl Iden for Metrics {
    fn unquoted(&self, s: &mut dyn std::fmt::Write) {
        write!(
            s,
            "{}",
            match self {
                Self::TableName(name) => name,
                Self::Timestamp => "timestamp",
                Self::MetricId => "metric_id",
                Self::Account => "account",
                Self::IntervalMs => "interval_ms",
                Self::Counter => "counter",
                Self::Gauge => "gauge",
                Self::MetricTagIds => "metric_tag_ids",
            }
        )
        .unwrap();
    }
}

fn main() {
    let select = Query::select()
        .columns([Metrics::Account, Metrics::Gauge])
        .from(Metrics::TableName("metric_1".to_string()))
        .build(PostgresQueryBuilder);
    println!("{}", select.0);
}
```

**Reply by tyt2y3** [reply]
I think there is a `Copy` constraint, so only `&'static str` or `usize` or `ShortString` can be used. But yes your idea is right. It's the reason why `Iden` trait does not simply return `&str`.

**Reply by theelderbeever** [reply]
@tyt2y3 the last example I sent compiles and works fwiw. Not sure if I will run into problems later?

**Reply by penso** [reply]
That's a nice short and concise example, I wonder how it'd work for migrations when I want to write a single migration but have, let's say, 10 of the same tables I need to run migrations on. In your scenario that'd be `account` and `gauge` when you create tables through migration files.

**Reply by penso** [reply]
Also as for my need, I use `find()` *everywhere*, rewriting those into sea-query would be such a pain

**Comment by tyt2y3** [comment]
Oh wow, that’s great news. Not really, once it’s Arced it’ll work as usual
________________________________

**Comment by BB-fat** [comment]
In my case the table name was calculated dynamically at runtime.
I wish there was a usage like this:
```rust
Entity::from_table("my_table").find()
```

**Comment by canxin121** [comment]
Is there any progress?

**Comment by penso** [comment]
I spent a bit of time on it, but I feel like there will be a lot of missing bits, feedback and help welcome: https://github.com/penso/sea-orm/pull/1
